### PR TITLE
fix filename pattern matching

### LINKED ISSUE #117: [Bug]: Starlark formatter no longer touches BUILD files
### What happened?

Looks like when format.sh was overhauled, rules_lint stopped getting BUILD/BUILD.bazel etc files outside of the root. the patterns for starlark at least are incomplete, and probably for other languages, too:

```console
$ git ls-files --cached --modified --other --exclude-standard BUILD.bazel
BUILD.bazel
$ find . -name BUILD.bazel | wc -l
737
```

looks like a little additional processing is needed from linguist to add '**/<PATTERN>' for any pattern that has no wildcards in it already

### Version

Development (host) and target OS/architectures: all

Output of `bazel --version`: bazel 7.0.0

Version of the Aspect rules, or other relevant rules from your
`WORKSPACE` or `MODULE.bazel` file:

Language(s) and/or frameworks involved: all

### How to reproduce

```shell
make an edit to a BUILD.bazel file in a subdirectory that would be corrected by buildifier. Run the formatter. Note it was not corrected.
```

### Any other information?

additionally, if you trace execution of ls-files in format.sh it looks like we're getting some unexpected wildcard handling:

```
...
++ case "$language" in
++ patterns=('BUCK' 'BUILD' 'BUILD.bazel' 'MODULE.bazel' 'Tiltfile' 'WORKSPACE' 'WORKSPACE.bazel' '*.bzl' '*.star')
++ '[' 0 -eq 0 ']'
++ git ls-files --cached --modified --other --exclude-standard BUCK BUILD BUILD.bazel MODULE.bazel Tiltfile WORKSPACE WORKSPACE.bazel extensions.bzl repositories.bzl '*.star'
...
```

notable here is the expansion of *.bzl into the two matching files in the root of my repository. it looks like this is because of an unquoted bash array expansion- might want to turn on shellcheck from rules_lint in rules_lint :)

### PR DESCRIPTION
**Pull request by pjjw**
git ls-files seems to match against filename strings exactly, and doesn't use the same logic as e.g. gitignore. this means that to match files by exact filename below the root, we need a */ at the start of the pattern to match more than files strictly at the root of the tree. these prefixed patterns however miss the root, so we need to include both forms.

additionally, the pattern array needs to be quoted to prevent the shell from doing glob expansion against any files in the current directory.

Fixes #117 and #122.

---

### Type of change

- Bug fix (change which fixes an issue)

### CODE REVIEW

**Comment by IamXander** [comment]
I totally missed this before I submitted https://github.com/aspect-build/rules_lint/issues/122 and https://github.com/aspect-build/rules_lint/pull/124. Both of these fix the same problem but in https://github.com/aspect-build/rules_lint/pull/124 might be a little nicer since you don't have to duplicate globs to include a "**" version.

**Comment by alexeagle** [comment]
Thanks both of you, this is my fault for not having more tests in place for the formatter before #105 . I'm going to work on adding those first so we don't keep chasing regressions.

**Comment by pjjw** [comment]
@IamXander the problem isn't fully fixed by #124, though- failing to recurse for e.g. `BUILD` files isn't caused by inadvertent glob expansion, but the (kinda inconsistent) semantics of `git ls-files`: if there's no wildcard, it's exact-match only.

File: format/private/format.sh [hunk 1]
```diff
@@ -54,15 +54,15 @@ function ls-files {
...
+
     if [ "$#" -eq 0 ]; then
         # When the formatter is run with no arguments, we run over "all files in the repo".
         # However, we want to ignore anything that is in .gitignore, is marked for delete, etc.
         # So we use `git ls-files` with some additional care.
 
         # TODO: determine which staged changes we should format; avoid formatting unstaged changes
         # TODO: try to format only modified regions of the file (where supported)
-        git ls-files --cached --modified --other --exclude-standard ${patterns[@]} | {
+        git ls-files --cached --modified --other --exclude-standard "${patterns[@]}" | {
```
Comment (pjjw) [inline format/private/format.sh:65]: @IamXander adding in a ` | tee /dev/stderr ` here is helpful for debugging, doing the same in #124 will show that even there only the root BUILD file is being considered without the add'l pattern expansion

**Comment by alexeagle** [comment]
Replaced by #127

**Pull request closed without merging**